codeholic

```diff
@@ -0,0 +1,180 @@
+    foreach ($args as $arg) {
+        if (strpos($arg, '=')) {
+            $newArgsArray[strtok($arg, '=')] = strtok('');
+            continue;
+        }
+
+        $newArgsArray[$arg] = false;
+    }
+
+    return $newArgsArray;
+}
+
+/**
+ * Serve the app using the command 'php codeholic serve'
+ * 
+ * @param array $argsArray => A restructered array containing command-line arguments
+ * 
+ * @return null
+ */
+function serveTheApp(array $argsArray)
+{
+    $pathToMainIndex    =   'public/';      // Path to the entry point of our app
+    $host               =   '127.0.0.1';    // Default 'host' url for our app
+    $port               =   8080;           // Default port for our app to use
+    $maxPort            =   null;           // Initial value of the '--max-port'          
+
+    // Possible parameters for the 'php codeholic serve' command
+    // For example, '--port=8080', '--host=127.0.0.1', '--max-port=9010'
+    $possibleParams = [
+        'host'      =>  '--host',
+        'port'      =>  '--port',
+        'maxPort'   =>  '--max-port'
+    ];
+
+    // If an invalid parameter is encountered, throw the exception
+    foreach ($argsArray as $arg => $value) {
+        if (!in_array($arg, $possibleParams)) {
+            throw new Exception("!!! 'php codeholic serve': Invalid parameter [ $arg ] !!!");
+        }
+
+        // If no value for parameter is passed, throw an exception
+        // In future, this needs to be changed, if we introduce a new parameter that does not require a value
+        if (!$value) {
+            throw new Exception("!!! 'php codeholic serve': Invalid value for the parameter [ $arg ] !!!");
+        }
+    }
+
+    // If the user provides a custom host url
+    if (array_key_exists($possibleParams['host'], $argsArray)) {
+        $host = $argsArray[$possibleParams['host']];
+    }
+
+    // If the user provides a custom port number
+    if (array_key_exists($possibleParams['port'], $argsArray)) {
+        $port = $argsArray[$possibleParams['port']];
+    }
+
+    // If the user explicitly specifies max port limit
+    if (array_key_exists($possibleParams['maxPort'], $argsArray)) {
+        $maxPort = $argsArray[$possibleParams['maxPort']];
+    }
+
+    // Validate port & max port as integers
+    validateInteger($port, "!!! 'php codeholic serve': Invalid argument value [ --port=$port ] !!!");
+
+    // Validate '--max-port' only when its value is provided
+    if ($maxPort) {
+        validateInteger((int) $maxPort, "!!! 'php codeholic serve': Invalid argument value [ --max-port=$maxPort ] !!!");
+    }
+
+    // Make sure '--max-port' is always greater than '--port'                  
+    if ($maxPort && $maxPort < $port) {
+        throw new Exception("!!! 'php codeholic serve': The value '--max-port=$maxPort' must not be less than '--port=$port' !!!");
+    }
+
+    // If default port is free, run PHP's built-in server on it
+    if (exec(`php -S $host:$port -t $pathToMainIndex`)) {
```

Comment:
When the default port is occupied and this code is executed. Code throws an error and does not try to allocate next port.

Reply:
@thecodeholic Thanks for notifying me about these issues. I'll try to fix them ASAP.